### PR TITLE
Revert "dbus: Switch default implementation to broker"

### PR DESCRIPTION
**Pull request by flokli**
This reverts commit e827697fd3435b458a759ef0b92de7f26624b910.

Supersedes #305273, which is drafted and with merge conflicts.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Aleksanaa** [comment]
Also the systemd team should be added to maintainers in dbus module later.

File: nixos/modules/services/system/dbus.nix
Comment (SuperSandro2000) [inline nixos/modules/services/system/dbus.nix:49]: What issues exactly? This vague comment isn't helpful for the reader.

I've been running this since months without any issues on all my systems, so it cannot be that experimental.

**Review by SuperSandro2000 — COMMENTED** [review]
We need a motivation here. 

I am guessing https://github.com/NixOS/nixpkgs/issues/297756#issuecomment-2082402329 ?

But there exists a comment that the issue is fixed with dbus-broker. nsspath is also not properly updated and there is another issue, which we cannot reproduce and maybe had something to do with nsspath.

**Comment by flokli** [comment]
The motivation is https://github.com/NixOS/nixpkgs/issues/302771, which is probably also the reason why @peterhoeg prepared the revert in https://github.com/NixOS/nixpkgs/pull/305273. @Aleksanaa also seems to be in agreement with that reasoning.

File: nixos/modules/services/system/dbus.nix
Comment (flokli) [reply nixos/modules/services/system/dbus.nix:49]: I can also entirely drop this line (it wasn't part of the revert per se). We can create a tracking issue to re-enable it past the next branchoff.

Comment (flokli) [reply nixos/modules/services/system/dbus.nix:49]: Done.

**Comment by alyssais** [comment]
Please include motivation for the revert in the commit message.

**Comment by flokli** [comment]
Done.

**Review by AndersonTorres — APPROVED** [review]

**Review by wegank — APPROVED** [review]

**Comment by magneticflux-** [comment]
> I fully support the temporary reversal of this change. However, unstable users who rely on automatic updates and remote updates may be a little unlucky 😋

Hi, it's me, unlucky :sweat_smile:
Activating a new NixOS profile (via `deploy-rs`) freezes after stopping `dbus-broker.service` and I eventually get dropped into a blank terminal.

For future reference, setting `services.dbus.implementation = "broker";` lets me activate a profile. The `--boot` flag (to update the boot loader without activating a profile right now) in `deploy-rs` also works after a reboot.

**Comment by flokli** [comment]
> > I fully support the temporary reversal of this change. However, unstable users who rely on automatic updates and remote updates may be a little unlucky 😋
> 
> Hi, it's me, unlucky 😅 Activating a new NixOS profile (via `deploy-rs`) freezes after stopping `dbus-broker.service` and I eventually get dropped into a blank terminal.

Did you encounter the same issue while the first switch (from dbus to dbus-broker) happened, too? Do you encounter it today, when switching from dbus to dbus-broker?

**Comment by arianvp** [comment]
Both dbus and dbus-broker are unrestartable. Normally  our switch-to-configuration takes care of not restarting them. But that logic is missed if you swap out the backend. 


Switching into a configuration that switches from one to the other is not supported and can not be supported.   We can not make it work.  There isn't much debugging we can do as it's not a feature that dbus supports.

If you still have problems after rebooting let us now

**Comment by SuperSandro2000** [comment]
So by reverting we are forcing everyone to reboot again, if I understand this correct?

**Comment by SuperSandro2000** [comment]
> The motivation is #302771

That's just a single person and no one could reproduce the issue and we have no logs. If it really boils down to some impermenance issue, I would not put the burden to fix highly experimental setups onto nixos.

**Comment by wegank** [comment]
> So by reverting we are forcing everyone to reboot again, if I understand this correct?

Yes. I guess there is no other way, since the systemd team has the final say on whether https://github.com/NixOS/nixpkgs/commit/e827697fd3435b458a759ef0b92de7f26624b910 will be in 24.05, and the proposal to [revert on release-24.05 instead of master](https://github.com/NixOS/nixpkgs/pull/305273#issuecomment-2067158830) was rejected.

**Comment by flokli** [comment]
Well, there is an alias `dbus.service` in place, to theoretically prevent the exact symptoms here from happening, delaying the dbus implementation switch until the next reboot. Unfortunately our switching script didn't account for this and still restarted it...

This is all stuff that might have ended up being surfaced if this feature would have gotten some more review :shrug: 

In subsequent PRs we can ensure the switching experience gets less disruptive, to be ready when we change the default again, but IMHO a slight bit of pain for unstable users to do this revert now is better than shipping a not-so-much-reviewed feature to the upcoming short release.

**Comment by SuperSandro2000** [comment]
So if I want to repropose the change, I would need to take a look into nixos-rebuild and exclude dbus-broker like it is already done for dbus? Maybe I find time going that.

**Comment by flokli** [comment]
As I said, not opposed to the change in general, just feel like we should do a lot more testing and make sure the switch is not the fire hazard it is right now.

**Comment by oxalica** [comment]
> > > I fully support the temporary reversal of this change. However, unstable users who rely on automatic updates and remote updates may be a little unlucky 😋
> > 
> > 
> > Hi, it's me, unlucky 😅 Activating a new NixOS profile (via `deploy-rs`) freezes after stopping `dbus-broker.service` and I eventually get dropped into a blank terminal.
> 
> Did you encounter the same issue while the first switch (from dbus to dbus-broker) happened, too? Do you encounter it today, when switching from dbus to dbus-broker?

I also got a blackscreen after hitting `switch` now. By login in a new tty and `tmux attach`, I can confirm the activation script tried to stop `dbus-broker.service`. A reboot can fix it, but with the cost of losing several minutes of data in GUI apps. I did not run into any issue when upgrading from dbus to dbus-broker, which was landed in nixos-unstable for quite a while. I don't remember if I did a `nixos-rebuild switch` or a `boot` for that upgrade. ~~Ironically, the reversion gives another chance for previously lucky ones to become unlucky, since the motivation of this PR is to fix upgrading failures, should we revert this PR again for the same reason?~~

**Comment by Aleksanaa** [comment]
> since the motivation of this PR is to fix upgrading failures

No, their problems also appear after rebooting. Next time maybe we'll go with a `stateVersion`.

**Comment by SuperSandro2000** [comment]
On a friends machine loginctl broke which made the nixos-rebuild switch already fail

```
dbus-daemon[2845527]: [system] Activating via systemd: service name='org.freedesktop.login1' unit='dbus-org.freedesktop.login1.service' requested by ':1.22' (uid=0 pid=2847559 comm="reboot" label="kernel")
dbus-daemon[2845527]: [system] Failed to activate service 'org.freedesktop.login1': timed out (service_start_timeout=25000ms)
```

**Comment by SuperSandro2000** [comment]
I think we have a massive problem. Machines in the state switched from dbus-broker to dbus cannot be rebooted with `reboot` or `systemctl reboot` and need to be hard-reseted 💀

**Comment by arianvp** [comment]
Yes that's bad. I'm happy we caught it now though? I'd be pretty miffed if rolling back from 24.05 to 23.11 would break rebooting without any mention of that in the release notes. 

Do we have a good understanding of why switching from dbus to dbus-broker doesn't cause issues but the other way around does? Did we skip restarting dbus in the one direction?

**Comment by arianvp** [comment]
My suggestion would be:  we use the exact same unit file name (dbus.service) instead of using the aliasing.  IThat way we're sure nixos-rebuild `systemctl restart`s (or ignores) it instead of `systemctl stop dbus-broker; systemctl start dbus`.   

Maybe the alias stuff only works one direction?
So it saved us for dbus->dbus-broker  but not the other way around?

**Comment by flokli** [comment]
I'd like to understand what's going on. Ideally, the activation script should understand what aliases are, so we don't need to deviate from how this is called on other distros to workaround bugs in our activation script.

As initially wrote, I propose we write a VM test with 2 specializations, testing switching back and forth, and ensure it keeps the old implementation running in both cases. That should allow testing out the behavior, ideally fixing the activation script.

**Comment by K900** [comment]
Looking into it.

**Comment by SuperSandro2000** [comment]
please review #309068

**Comment by Lyndeno** [comment]
> > since the motivation of this PR is to fix upgrading failures
> 
> No, their problems also appear after rebooting. Next time maybe we'll go with a `stateVersion`.

> anything relying on polkit breaks

I was reverted back to dbus, polkit operations like pkexec and using polkit authentication with 1password do not work in Gnome, the authorization window does not pop up. Is this related?

I switched back to dbus-broker and am still experiencing the same issues

**Comment by arianvp** [comment]
Most likely yes. Do your problems persist across reboot? If so then this is probably another issue.

**Comment by Lyndeno** [comment]
> Most likely yes. Do your problems persist across reboot? If so then this is probably another issue.

Persists across reboot, I will see if #308884 fixes it. I see the same errors when running sudo or in the journal whenever polkit is attempted.

**Pull request merged**
